File: Python/ByRound/2065/2065_H_Bro_Thinks_He_s_Him.py
# Problem: CF 2065 H - Bro Thinks He's Him
# https://codeforces.com/contest/2065/problem/H

# ===================================================================
# Problem: H. Bro Thinks He's Him
# Purpose: For a binary string s, compute the sum of f(b) over all non-empty subsequences b,
#          where f(t) is the minimum number of contiguous identical substrings in t.
#          After each query flipping a bit in s, output the result modulo 998244353.
#
# Algorithm: Segment Tree with lazy propagation for efficient updates and range queries.
#            Each node stores four values (c00, c01, c10, c11) and (s00, s01, s10, s11)
#            representing counts and sums of subsequences ending at certain character types
#            during interval operations.
#
# Time Complexity: O(n log n + q log n)
# Space Complexity: O(n)
# ===================================================================

import sys

input = lambda: sys.stdin.readline().rstrip()
ii = lambda: int(input())
mii = lambda: map(int, input().split())
lii = lambda: list(mii())
MOD = 998244353


class SegmentTree:
    __slots__ = "height", "size", "c00", "c01", "c10", "c11", "s00", "s01", "s10", "s11"

    def __init__(self, nums):
        n = len(nums)
        self.height = (n - 1).bit_length()  # Height of segment tree
        self.size = 1 << self.height       # Size of the segment tree array
        self.c00 = [0] * (2 * self.size)    # Count values for transitions
        self.c01 = [0] * (2 * self.size)
        self.c10 = [0] * (2 * self.size)
        self.c11 = [0] * (2 * self.size)
        self.s00 = [0] * (2 * self.size)    # Sum of f values for subseqs
        self.s01 = [0] * (2 * self.size)
        self.s10 = [0] * (2 * self.size)
        self.s11 = [0] * (2 * self.size)
        # Initialize leaves with binary values
        for idx, val in enumerate(nums, self.size):
            if not val:
                self.c00[idx] = self.s00[idx] = 1
            else:
                self.c11[idx] = self.s11[idx] = 1
        # Build the segment tree from bottom up
        for idx in range(self.size - 1, 0, -1):
            self.pushup(idx)

    def pushup(self, rt):
        # Compute combined values for a node based on left and right child nodes
        ln, rn = rt * 2, rt * 2 + 1
        lc00, lc01, lc10, lc11, ls00, ls01, ls10, ls11 = (
            self.c00[ln],
            self.c01[ln],
            self.c10[ln],
            self.c11[ln],
            self.s00[ln],
            self.s01[ln],
            self.s10[ln],
            self.s11[ln],
        )
        rc00, rc01, rc10, rc11, rs00, rs01, rs10, rs11 = (
            self.c00[rn],
            self.c01[rn],
            self.c10[rn],
            self.c11[rn],
            self.s00[rn],
            self.s01[rn],
            self.s10[rn],
            self.s11[rn],
        )
        # Update count and sum values
        self.c00[rt] = ((lc00 + lc01) * (rc00 + rc10) + lc00 + rc00) % MOD
        self.c01[rt] = ((lc00 + lc01) * (rc01 + rc11) + lc01 + rc01) % MOD
        self.c10[rt] = ((lc10 + lc11) * (rc00 + rc10) + lc10 + rc10) % MOD
        self.c11[rt] = ((lc10 + lc11) * (rc01 + rc11) + lc11 + rc11) % MOD
        self.s00[rt] = (
            (ls00 + ls01) * (rc00 + rc10)
            + (lc00 + lc01) * (rs00 + rs10)
            - lc00 * rc00
            - lc01 * rc10
            + ls00
            + rs00
        ) % MOD
        self.s01[rt] = (
            (ls00 + ls01) * (rc01 + rc11)
            + (lc00 + lc01) * (rs01 + rs11)
            - lc00 * rc01
            - lc01 * rc11
            + ls01
            + rs01
        ) % MOD
        self.s10[rt] = (
            (ls10 + ls11) * (rc00 + rc10)
            + (lc10 + lc11) * (rs00 + rs10)
            - lc10 * rc00
            - lc11 * rc10
            + ls10
            + rs10
        ) % MOD
        self.s11[rt] = (
            (ls10 + ls11) * (rc01 + rc11)
            + (lc10 + lc11) * (rs01 + rs11)
            - lc10 * rc01
            - lc11 * rc11
            + ls11
            + rs11
        ) % MOD

    def inverse(self, idx):
        # Flip the bit at index idx and update affected nodes
        idx += self.size
        self.c00[idx] ^= 1
        self.c11[idx] ^= 1
        self.s00[idx] ^= 1
        self.s11[idx] ^= 1
        for i in range(1, self.height + 1):
            self.pushup(idx >> i)

    def all(self):
        # Return total sum of f over all subsequences
        return (self.s00[1] + self.s01[1] + self.s10[1] + self.s11[1]) % MOD


def solve():
    nums = list(map(int, input()))
    seg = SegmentTree(nums)
    nq = ii()
    ans = [0] * nq
    for qi, idx in enumerate(mii()):
        idx -= 1
        seg.inverse(idx)
        ans[qi] = seg.all()
    return ans


for _ in range(ii()):
    print(*solve())


# https://github.com/VaHiX/codeForces/

File: CPP/ByTier/H/2065_H_Bro_Thinks_He_s_Him.cpp
// Problem: CF 2065 H - Bro Thinks He's Him
// https://codeforces.com/contest/2065/problem/H

/*
 * Problem: H. Bro Thinks He's Him
 * Purpose: Given a binary string and a series of flips, compute the sum of f(b) 
 *          over all non-empty subsequences b of s, where f(t) is the number of 
 *          contiguous segments of identical characters in t.
 * 
 * Algorithm:
 * - Use Binary Indexed Trees (Fenwick Trees) to maintain prefix and suffix sums efficiently.
 * - Precompute powers of 2 modulo 998244353 for fast computation.
 * - For each query, update the BITs and calculate the change in the total sum based on 
 *   the contribution of the flipped bit to all subsequences.
 * 
 * Time Complexity: O((n + q) * log n) per test case, where n is the length of the string.
 * Space Complexity: O(n) for the BITs, power array, and other auxiliary arrays.
 */

#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

using namespace std;
using ll = long long;
const int MOD = 998244353;
const int MAXN = 200000 + 5;
static int fenw_pref[2][MAXN], fenw_suff[2][MAXN];
static int pow2[MAXN];

// Add value to BIT at index idx
inline void bit_add(int f[], int n, int idx, int val) {
  for (int i = idx; i <= n; i += i & -i) {
    int t = f[i] + val;
    if (t >= MOD)
      t -= MOD;
    f[i] = t;
  }
}

// Compute prefix sum up to index idx in BIT
inline int bit_sum(int f[], int idx) {
  int res = 0;
  for (int i = idx; i > 0; i -= i & -i) {
    res += f[i];
    if (res >= MOD)
      res -= MOD;
  }
  return res;
}

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  
  // Precompute powers of 2
  pow2[0] = 1;
  for (int i = 1; i < MAXN; i++) {
    pow2[i] = (int)((ll)pow2[i - 1] * 2 % MOD);
  }
  
  int t;
  cin >> t;
  while (t--) {
    string s;
    cin >> s;
    int n = (int)s.size();
    int q;
    cin >> q;
    vector<int> v(q);
    for (int i = 0; i < q; i++) {
      cin >> v[i];
    }
    
    // Initialize the BITs
    for (int c = 0; c < 2; c++) {
      for (int i = 1; i <= n; i++) {
        fenw_pref[c][i] = 0;
        fenw_suff[c][i] = 0;
      }
    }
    
    ll S1 = (ll)n * pow2[n - 1] % MOD; // Base contribution from all subsequences
    ll S2 = 0; // Dynamic contribution from individual prefix/suffix changes
    
    int sum_pow2_pref[2] = {0, 0};
    int total_suff[2] = {0, 0};
    
    // Initialize initial BITs and compute S2
    for (int j = 1; j <= n; j++) {
      int f = s[j - 1] - '0';
      S2 += (ll)sum_pow2_pref[f] * pow2[n - j] % MOD;
      if (S2 >= MOD)
        S2 -= MOD;
      sum_pow2_pref[f] += pow2[j - 1];
      if (sum_pow2_pref[f] >= MOD)
        sum_pow2_pref[f] -= MOD;
      bit_add(fenw_pref[f], n, j, pow2[j - 1]);
      bit_add(fenw_suff[f], n, j, pow2[n - j]);
      total_suff[f] = (total_suff[f] + pow2[n - j]) % MOD;
    }
    
    // Process queries
    for (int qi = 0; qi < q; qi++) {
      int pos = v[qi];
      int f_old = s[pos - 1] - '0';
      int f_new = 1 - f_old;
      
      // Calculate contributions before and after flip
      int sum_pref_old = bit_sum(fenw_pref[f_old], pos - 1);
      int sum_pref_new = bit_sum(fenw_pref[f_new], pos - 1);
      int pref_suff_old = bit_sum(fenw_suff[f_old], pos);
      int pref_suff_new = bit_sum(fenw_suff[f_new], pos);
      
      int sum_suff_old = total_suff[f_old] - pref_suff_old;
      if (sum_suff_old < 0)
        sum_suff_old += MOD;
      int sum_suff_new = total_suff[f_new] - pref_suff_new;
      if (sum_suff_new < 0)
        sum_suff_new += MOD;
      
      // Compute the change in S2
      int d1 = sum_pref_new - sum_pref_old;
      if (d1 < 0)
        d1 += MOD;
      int d2 = sum_suff_new - sum_suff_old;
      if (d2 < 0)
        d2 += MOD;
      
      ll delta = ((ll)d1 * pow2[n - pos] + (ll)d2 * pow2[pos - 1]) % MOD;
      S2 += delta;
      if (S2 >= MOD)
        S2 -= MOD;
      
      // Update BITs and total suff sums
      bit_add(fenw_pref[f_old], n, pos, (MOD - pow2[pos - 1]) % MOD);
      bit_add(fenw_suff[f_old], n, pos, (MOD - pow2[n - pos]) % MOD);
      total_suff[f_old] = (total_suff[f_old] - pow2[n - pos] + MOD) % MOD;
      bit_add(fenw_pref[f_new], n, pos, pow2[pos - 1]);
      bit_add(fenw_suff[f_new], n, pos, pow2[n - pos]);
      total_suff[f_new] = (total_suff[f_new] + pow2[n - pos]) % MOD;
      
      // Flip the character in s
      s[pos - 1] = char('0' + f_new);
      
      // Output the final answer for the current query
      int ans = (int)(S1 - S2);
      if (ans < 0)
        ans += MOD;
      cout << ans << (qi + 1 < q ? ' ' : '\n');
    }
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/